### PR TITLE
Enable auto_num_workers

### PR DESCRIPTION
**Pull request by eunwoosh**
### Summary
Implement auto_num_workers.
It simply decides proper num_workers values based on number of GPU and number of CPU cores.
Additionally, I change HPO test code to check all ckpt files are removed except best wegith.
It's because model weights of last trials tried right before HPO is done can be left if HPO algo is finished before trials is finalized.

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/core/data/module.py [hunk 1]
```diff
@@ -65,6 +67,16 @@ def __init__(
...
 
+        if self.config.auto_num_workers:
+            if self.config.device not in [DeviceType.gpu, DeviceType.auto]:
+                log.warning(
+                    "Only GPU device type support auto_num_workers. "
+                    f"Current deveice type is {self.config.device!s}. auto_num_workers is skipped.",
+                )
+            elif (num_workers := get_adaptive_num_workers()) is not None:
+                for subset_config in config_mapping.values():
+                    subset_config.num_workers = num_workers
```
Comment (vinnamkim) [inline src/otx/core/data/module.py:82]: Add info message to notify an user that `num_workers` changed by `auto_num_workers=True`.

Comment (eunwoosh) [reply src/otx/core/data/module.py:82]: done at [3c84d46](https://github.com/openvinotoolkit/training_extensions/pull/2931/commits/3c84d46afb26d08a5d81469a9c502007e72ed61c)

**Review by vinnamkim — DISMISSED** [review]

**Review by jaegukhyun — DISMISSED** [review]

**Review by sungmanc — APPROVED** [review]

**Review by goodsong81 — COMMENTED** [review]
Is it all working well on CPU-only system or CPU version of pytorch?
(I suppose so, but there was many issues in 1.x)

**Review by jaegukhyun — APPROVED** [review]

**Comment by eunwoosh** [comment]
> Is it all working well on CPU-only system or CPU version of pytorch? (I suppose so, but there was many issues in 1.x)

Yes because if `torch.cuda.device_count()` return 0, nothing won't executed.

**Review by harimkang — APPROVED** [review]

**Pull request merged**